### PR TITLE
Firo full node wallet support

### PR DESCRIPTION
**Pull request by dev-warrior777**
This is a draft showing compatibility using `dcrdex/dex/networks/firo/params_test.go` with P2PKH and P2SH txs generated using `jrick` atomicswap code for Zcoin (Firo)

### CODE REVIEW

**Comment by dev-warrior777** [comment]
The above is for communication with no expectation of being merged in it's current state.

Next we want to make a client wallet. What are the most useful golang tests recommended for this?
As noted Firo does not have `creatwallet` RPC so tmux test harness will be a bit more challenging

**Comment by dev-warrior777** [comment]
`panic: failed to register firo parameters: duplicate Bitcoin network`

Any clues for this would be helpful. Thanks in advance!

`dcrdex/dex/networks/firo/params_test.go` run by itself always passes on mainnet and testnet

**Comment by chappjc** [comment]
Every `btc.CloneParams.Net` must be unique.  I haven't checked yours, but assuming you've update the mainnet and testnet ones properly, you probably just need to tweak the regnet value.

**Comment by dev-warrior777** [comment]
You were correct many of the old coins copy the same bitcoin network bytes. Thanks for the good clue.

**Comment by dev-warrior777** [comment]
So learning how stuff works and nicely integrating Firo code harnesses and testing while the RPC and other code is changing fast underneath me is .. e.g. bucks's CallPPC ... it is all getting away from me somewhat with the base I was working from. Wen stable code ;-)

Let's have a quick review on the above commits -- then I will close this PR and rework the above with and updated code base

Thanks!

File: dex/testing/firo/harness.sh [hunk 1]
```diff
@@ -0,0 +1,331 @@
+#!/usr/bin/env bash
+SYMBOL="firo"
+FIRO_BIN_DIR="~/firo/bin"
+DAEMON="${FIRO_BIN_DIR}/firod"
+CLI="${FIRO_BIN_DIR}/firo-cli"
```
Comment (buck54321) [inline dex/testing/firo/harness.sh:5]: These'll just need to be expected in `PATH`.

**Comment by buck54321** [comment]
I had to apply the following changes to get things working.

https://github.com/buck54321/dcrdex/compare/b89e1b9...buck54321:dcrdex:firo-fixes

I built **firod** on its master branch.

File: dex/testing/firo/harness.sh (hunk 1 above)
Comment (dev-warrior777) [reply dex/testing/firo/harness.sh]: Removed extra path variables in [119a661](https://github.com/decred/dcrdex/pull/2270/commits/119a661949de095ef6b7dc7ded2a1f2e55bbd585) below.

**Comment by dev-warrior777** [comment]
Thanks for the help bucks ...

**Comment by dev-warrior777** [comment]
actually . . why don't we just merge this?

**Comment by dev-warrior777** [comment]
Resolved merge conflicts so `dgb` and `firo` can coexist

**Comment by chappjc** [comment]
@dev-warrior777 we never want to merge master into feature branches.  The preferred workflow is to rebase after you update master.  Please squash away the merge commit.  Thanks.

**Comment by dev-warrior777** [comment]
> @dev-warrior777 we never want to merge master into feature branches. The preferred workflow is to rebase after you update master. Please squash away the merge commit. Thanks.

I am not great with git .. can you advise how to do this on my local, pls.

**Comment by chappjc** [comment]
Will get to you re: git in a bit.  Killing a merge commit generally requires a hard reset and force push.

**Comment by dev-warrior777** [comment]
> Will get to you re: git in a bit. Killing a merge commit generally requires a hard reset and force push.

Okie doke .. understand .. let's get 0.6 out of the door. Thanks.

**Comment by dev-warrior777** [comment]
> Will get to you re: git in a bit. Killing a merge commit generally requires a hard reset and force push.

Did that: Looks like we are back to here ` 24598c1 ` above ..

image

**Comment by dev-warrior777** [comment]
Rebased firoX

**Comment by dev-warrior777** [comment]
```
Simnet Trade tests for pair: dcrfiro

success		SUCCESS
nomakerswap	SUCCESS
notakerswap	ERROR: test "notakerswap" failed with client 1 as maker: 
                       'client 1 reported 1 unrefunded swaps after 30s'

			Solution.
			-----------
			DCR as Maker:
			"Mined 6 blocks for assetID 42 to expire swap locktimes"
			Test 1. 
			**Manually mine DCR** from harness a few times while waiting
			for the refund. 
			Test 2. 
			Firo mining is ok for the second refund
			swap test.
                       
                        SUCCESS

nomakerredeem	ERROR: test "nomakerredeem" failed with client 1 as maker:
			'client 1 reported 1 unrefunded swaps after 30s'

			Solution.
			-----------
			Test 1.
			"Mined 6 blocks for assetID 42 to expire swap locktimes"
			**Manually mine DCR** from harness a few times ...
			Test 2.
			"Mined 6 blocks for assetID 42 to expire swap locktimes"
			**Manually mine DCR** from harness a few times ...
                     
                        SUCCESS

makerghost	SUCCESS
orderstatus	SUCCESS
resendpending	SUCCESS
-------------------------------------
```

**Comment by chappjc** [comment]
> notakerswap	ERROR: test "notakerswap" failed with client 1 as maker: 
>                        'client 1 reported 1 unrefunded swaps after 30s'

Will have a looks, but since the "miner" tmux windows were recently added, I think it interferes with the intended block cadence in these simnet trade tests.  Did you have the miner `watch` commands running or did you stop them first?  The simnet trade tests will mine blocks as needed.

**Comment by dev-warrior777** [comment]
> Did you have the miner watch commands running or did you stop them first? The simnet trade tests will mine blocks as needed.

I did have the constant 15s mining running. Now I switched it off I have to manually mine DCR for the fee payment setup (both coins) plus other mining issues. Maybe the dcr harness and simnet-trade-tests/simnet_trading.go out of sync with the harness somehow. Now tests I previously had working above - the basic success path - are not working.

Switching the watch mining back on again I get the results documented above reliably yet again.

Anyway not so important as I feel the latest Firo code is now proven for all the Simnet tests -- particularly on the refund paths which were previously failing due to the need to manually mine. So we can have another review ;-)

**Comment by chappjc** [comment]
Yes, the Firo bits look fine.  I'm just trying to clarify that the simnet trade tests are meant to be run without the `watch`es mining periodically.

Not sure why it fails to work at all for you without the miner watches running.  I just now turned off the miners for both dcr and btc (you have to change tmux windows to find the btc one), and registration worked fine and the success tests passed too.  Going to check with firo shortly.

File: client/cmd/dexcctl/simnet-setup.sh [hunk 2]
```diff
@@ -61,6 +64,11 @@ if [ $DOGE_ON -eq 0 ]; then
 	./dexcctl -p abc -p "" --simnet newwallet 3 dogecoindRPC ~/dextest/doge/alpha/alpha.conf
 fi
 
+if [ $FIRO_ON -eq 0 ]; then
+	echo configuring firo wallet
+	./dexcctl -p abc -p "" --simnet newwallet 3 firodRPC ~/dextest/firo/alpha/alpha.conf
```
Comment (chappjc) [inline client/cmd/dexcctl/simnet-setup.sh:69]: The "3" here is where the asset ID goes, so `136`

File: client/asset/btc/btc.go [hunk 3]
```diff
@@ -976,6 +976,11 @@ func (btc *ExchangeWalletSPV) RemovePeer(addr string) error {
 func (btc *ExchangeWalletFullNode) FeeRate() uint64 {
 	// NOTE: With baseWallet having an optional external fee rate source, we may
 	// consider making baseWallet a FeeRater by allowing a nil local func.
+	if btc.walletInfo.Name == "Firo" {
+		// estimatesmartfee unstable for firo, estimatefee not much better ;(
```
Comment (chappjc) [inline client/asset/btc/btc.go:980]: Is this a widely-known or documented issue?  For most assets, btc included, estimatesmartfee doesn't want to work unless the chain is well utilized and the node has warmed up for over an hour.  This means that these RPCs typically don't work on simnet, often not on testnet either.

File: client/asset/btc/btc.go [hunk 4]
```diff
@@ -4246,17 +4251,89 @@ func (btc *baseWallet) DepositAddress() (string, error) {
...
 	return addrStr, nil
 }
 
+// getPrivKeyForAddress accomodates `dumpprivkey' RPC for different RPC sequencing
+func (btc *baseWallet) getPrivKeyForAddress(addr string) (*btcec.PrivateKey, error) {
+	if btc.walletInfo.Name == "Firo" {
+		return btc.getPrivkeyFiro(addr)
+	}
+	return btc.node.privKeyForAddress(addr)
+}
```
Comment (chappjc) [inline client/asset/btc/btc.go:4272]: I think the preferred way to override functionality is like this: https://github.com/decred/dcrdex/commit/612e57b7a31093e7f23bf7c926d74a9bfc92ab4c
At least this is the direction @buck54321 was going with the last zec update, so maybe he can confirm.

**Review by chappjc — CHANGES_REQUESTED** [review]

**Comment by dev-warrior777** [comment]
Sorry just woke up .. here is 3am .. thanks a lot for reviewing I will look at the details tomorrow but I like the bucks idea to use a function pointer for the weird privkey rpc firo has. On `estimatesmartfee` .. even the c++ rpc client help documents  this rpc as unstable and subject to change.

File: client/asset/btc/btc.go (hunk 3 above)
Comment (chappjc) [reply client/asset/btc/btc.go:980]: Related, looks like we can use "https://explorer.firo.org/insight-api-zcoin/utils/estimatefee" for an `ExternalFeeEstimator`

File: client/asset/btc/livetest/livetest.go [hunk 5]
```diff
@@ -118,6 +118,8 @@ func (rig *testRig) mineAlpha() error {
 	tmuxWindow := rig.symbol + "-harness:2"
 	if rig.symbol == "zec" {
 		tmuxWindow = rig.symbol + "-harness:4"
+	} else if rig.symbol == "firo" {
+		tmuxWindow = rig.symbol + "-harness:4"
 	}
```
Comment (chappjc) [inline client/asset/btc/livetest/livetest.go:123]: Looks like "doge" is also a window-4 tmux setup that we missed.  Let's convert this to a `switch` as follows;

```go
var tmuxWindow string
switch rig.symbol {
case "zec", "firo", "doge":
	tmuxWindow = rig.symbol + "-harness:4"
default:
	tmuxWindow = rig.symbol + "-harness:2"
}
```

File: client/cmd/simnet-trade-tests/run [hunk 6]
```diff
@@ -86,6 +90,7 @@ ltcelectrumdcr - Litecoin Electrum wallet and Decred RPC wallet on LTC-DCR marke
 dcrdoge - RPC wallets on DCR-DOGE market
 dcrdgb - RPC wallets on DCR-DGB market
 dcreth - Decred RPC wallet and Ethereum native wallet on DCR-ETH market
+dcrfiro - RPC wallets on DCR-FIRO market
```
Comment (chappjc) [inline client/cmd/simnet-trade-tests/run:93]: These all passed for me with the miner tmux watches stopped.

File: client/webserver/site/src/js/doc.ts [hunk 7]
```diff
@@ -20,6 +20,7 @@ const BipIDs: Record<number, string> = {
   3: 'doge',
   145: 'bch',
   60: 'eth',
+  136: 'firo',
```
Comment (chappjc) [inline client/webserver/site/src/js/doc.ts:23]: Please resize the firo.png to 100x100 and run it through optipng, and firo.svg through svgo.

File: client/webserver/site/src/js/order.ts [hunk 8]
```diff
@@ -567,6 +567,10 @@ const CoinExplorers: Record<number, Record<number, (cid: string) => string>> = {
     [Mainnet]: (cid: string) => `https://dogeblocks.com/tx/${cid.split(':')[0]}`,
     [Testnet]: (cid: string) => `https://blockexplorer.one/dogecoin/testnet/tx/${cid.split(':')[0]}`
   },
+  136: { // firo
+    [Mainnet]: (cid: string) => `https://explorer.firo.org/tx/${cid.split(':')[0]}`,
+    [Testnet]: (cid: string) => `https://testexplorer.firo.org/tx/${cid.split(':')[0]}`
```
Comment (chappjc) [inline client/webserver/site/src/js/order.ts:572]: https://testexplorer.firo.org/ seesm to be stuck for a few days.  Any way to ping the admins?

**Review by chappjc — CHANGES_REQUESTED** [review]

File: dex/testing/firo/harness.sh [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+mkdir ${NODES_ROOT}/\$1
+
+printf "rpcuser=user\nrpcpassword=pass\ndatadir=${NODES_ROOT}/\$1\ntxindex=1\nregtest=1\ndandelion=0\nrpcport=\$2\n" > ${NODES_ROOT}/\$1/\$1.conf
+
+${DAEMON} -rpcuser=user -rpcpassword=pass \
+-rpcport=\$2 -datadir=${NODES_ROOT}/\$1 -txindex=1 -regtest=1 -dandelion=0 \
+-debug=rpc -debug=net -debug=mempool -debug=walletdb -debug=addrman -debug=mempoolrej \
+-whitelist=127.0.0.0/8 -whitelist=::1 \
+-port=\$3 -fallbackfee=0.00001 -daemon  # -daemon .. if you want to stop it with stop-wallet
```
[the remark this answers is not in the document]
Comment (dev-warrior777) [reply dex/testing/firo/harness.sh:210]: I tried `estimatefee` on mainnet explorer https://explorer.firo.org/insight-api-zcoin/utils/estimatefee
At least it has a history of Txs there and it returns: 0.00000999
So I think can be used.
------------------------------

`-server` is better in the harness, updated and will push

File: client/asset/btc/livetest/livetest.go (hunk 5 above)
Comment (dev-warrior777) [reply client/asset/btc/livetest/livetest.go:123]: Better and future-proof, thanks
Updated and will push

File: client/cmd/dexcctl/simnet-setup.sh (hunk 2 above)
Comment (dev-warrior777) [reply client/cmd/dexcctl/simnet-setup.sh:69]: Updated thanks, will push

**Comment by dev-warrior777** [comment]
Updated firo client with almost all of the requested review changes. 
I still owe you the resized png/svg images. And I want to look further at fee estimation. For sure we cannot use `estimatesmartfee` but server side can probably use `estimatefee` .. we already specify `BackendCloneConfig.DumbFeeEstimates: true` in `server/asset/firo/firo.go`

**Comment by chappjc** [comment]
> server side can probably use `estimatefee` .. we already specify `BackendCloneConfig.DumbFeeEstimates: true` in `server/asset/firo/firo.go`

Thanks.  `estimatefee` worked in my testing.  server-side `DumbFeeEstimates` option is not desirable; ideally just a fallback.
Client side we can also use `estimatefee`, plus we can define `ExternalFeeEstimator` (client's fallback) to use https://explorer.firo.org/insight-api-zcoin/utils/estimatefee.  Use DOGE's impl for an example.  It's very simple.

File: client/webserver/site/src/js/doc.ts (hunk 7 above)
Comment (dev-warrior777) [reply client/webserver/site/src/js/doc.ts:23]: As requested .. fixed in https://github.com/decred/dcrdex/pull/2270/commits/e111e14029f7062e554ebb3fe0e59490bd98e8d7

File: client/asset/btc/btc.go (hunk 3 above)
Comment (dev-warrior777) [reply client/asset/btc/btc.go]: Updates as discussed: in https://github.com/decred/dcrdex/pull/2270/commits/cb14f9a01722cc7a0e62d4def5971b7bb8767fc9

File: client/cmd/simnet-trade-tests/run (hunk 6 above)
Comment (dev-warrior777) [reply client/cmd/simnet-trade-tests/run:93]: > These all passed for me with the miner tmux watches stopped.

Me too now .. my issue was an environment issue when running `simnet-trade-tests` binary itself.

**Comment by dev-warrior777** [comment]
Simnet tests passing now with watch mining off .. my issue was an environment variable: PATH when running simnet-trade-tests binary itself.

File: client/webserver/site/src/js/order.ts (hunk 8 above)
Comment (dev-warrior777) [reply client/webserver/site/src/js/order.ts:572]: At the moment you can just cpu mine testnet. 
Anwar in https://t.me/firosupport can gpu mine it when it gets busy. UTC+8 is his timezone (as is mine)

Comment (chappjc) [reply client/webserver/site/src/js/order.ts:572]: The Firo community should keep it moving

Comment (dev-warrior777) [reply client/webserver/site/src/js/order.ts:572]: Yes they should .. actually will get busier this month with some release candidates of Lelantus Spark protocol which community will help test.
There is also a semi-private `devnet` where most of the zkp changes get developed  which is more used than this testnet. 

For me I could generate blocks on testnet easily this week. 
`./firod -testnet -daemon`
...
`./firo-cli -testnet generate 1` 
will move the explorer 1 block
Sync: This testnet current size is maybe 200MB

and if want coins there is a testnet faucet at:
https://testexplorer.firo.org/faucet
..with 200k coins

**Comment by buck54321** [comment]
> Thanks. `estimatefee` worked in my testing (with 2-6 blocks as target, not 1). server-side `DumbFeeEstimates` option is not desirable; ideally just a fallback.

`DumbFeeEstimates` isn't that bad. It's `ManualMedianFee` that we really want to avoid.

**Comment by chappjc** [comment]
> > Thanks. `estimatefee` worked in my testing (with 2-6 blocks as target, not 1). server-side `DumbFeeEstimates` option is not desirable; ideally just a fallback.
> 
> `DumbFeeEstimates` isn't that bad. It's `ManualMedianFee` that we really want to avoid.

Right, good point, it's fine if there's getblockstats.

**Review by chappjc — APPROVED** [review]
Haven't retested with the fee estimator updates, but the simnet trade suite was passing, so. LGTM!

**Comment by dev-warrior777** [comment]
Ok, thanks for review. 
I am already doing a lot of prep for DCRDEX electrum-firo light wallet .. like fixing `regtest` functionality in both https://github.com/firoorg/electrumx-firo & https://github.com/firoorg/electrum-firo -- gonna put PR's in for those Firo repos today.

So if we can merge the full node support above I would be grateful. I can then work with a less stale tree for that.

**Comment by dev-warrior777** [comment]
> I am already doing a lot of prep for DCRDEX electrum-firo light wallet .. like fixing `regtest` functionality in both https://github.com/firoorg/electrumx-firo & https://github.com/firoorg/electrum-firo -- gonna put PR's in for those Firo repos today.

https://github.com/firoorg/electrumx-firo/pull/4
https://github.com/firoorg/electrum-firo/pull/41

**Review by buck54321 — APPROVED** [review]

**Pull request merged**